_posts/2020-05-05-churn-vs-complexity-vs-coverage.markdown

```diff
@@ -0,0 +1,255 @@
+- Is our application poorly designed?
+
+If your legacy application is written in Ruby, you can use a gem called [`churn`](https://github.com/danmayer/churn)
+to find the files that are constantly changing.
+
+<h3 id="calculating-churn">Calculating Churn in Ruby</h3>
+
+You can install [`churn`](https://rubygems.org/gems/churn) like this:
+
+    gem install churn
+
+Then call it like this:
+
+    churn -e rb --start_date "10 years ago" lib -i spec
+
+This will show you a list of files sorted by their churn count (number of times
+that the file has changed in the SCM that you are using)
+
+```
+**********************************************************************
+* Project Churn
+**********************************************************************
+
+Files
++-----------------------------------------------------------------------------------+---------------+
+| file_path                                                                         | times_changed |
++-----------------------------------------------------------------------------------+---------------+
+| lib/metric_fu.rb                                                                  | 50            |
+| lib/metric_fu/version.rb                                                          | 44            |
+| lib/metric_fu/configuration.rb                                                    | 36            |
+| lib/metric_fu/reporting/templates/awesome/awesome_template.rb                     | 29            |
+| lib/metric_fu/io.rb                                                               | 23            |
+| lib/metric_fu/formatter/html.rb                                                   | 23            |
+| lib/base/configuration.rb                                                         | 23            |
+| lib/metric_fu/run.rb                                                              | 22            |
+| lib/metric_fu/metrics/saikuro/saikuro.rb                                          | 20            |
+| lib/metric_fu/metric.rb                                                           | 20            |
++-----------------------------------------------------------------------------------+---------------+
+```
+
+By itself, this metric provides a limited insight:
+
+> These are the files that have changed the most since the beginning of the
+project
+
+It doesn't provide insights into why these files are changing as much as they
+do. If you open these files in your editor, you might find that files with high
+churn are super complex or super simple.
+
+<h2 id="complexity">Complexity</h2>
+
+As you go about making your changes, you will introduce complexity. There is no
+way around it. The best way to have no complexity is to have no code.
+
+It doesn't matter whether you are a rockstar object-oriented architect or a
+novice programmer. You will introduce complexity into a project.
+
+"Complexity" should not be considered a negative term. "Extreme complexity"
```

Comment:
Unwarranted or Unnecessary complexity is the anti-pattern here.  Sometimes something is just complicated by nature.